### ISSUE TITLE
DeleteCommit + New Commit/Job Model

### ISSUE THREAD

**Issue by msteffen** [pachyderm/pachyderm#2426]
# Objective
Change PFS's commit model and PPS's job model, with the aims of:
1. Simplifying Pachyderm's job-triggering semantics
2. Allowing users to delete past commits

# Background
Currently, when a user finishes a commit in PFS, PPS is responsible for observing the change, spawning any downstream jobs, and creating and finishing an output commit in any downstream pipeline's output repo. This has created two problems:

1. PPS must trace a pipeline's provenance all the way back the input repos<sup>1</sup> on which it depends and watch those for new commits. When a new input commit is created, PPS can't start a distant downstream job until all intermediate jobs succeed (so that it does no processing and creates no output commit if any of those intermediate jobs fail). If a pipeline's master is restarted while following the progress of a new input commit, it must reconstruct the status of all upstream input commits and jobs to before resuming, which we have struggled to implement correctly.

2. PFS is designed to operate as an independent system underneath PPS. If a user deletes a commit, PFS therefore has no way to stop all jobs that depend on that commit before actually deleting the commit's data, making  it an unsafe operation.

In our new design, all jobs will have an output commit. A job's output commit will be created by PFS even before the job itself is created in etcd. Pipeline masters will subscribe to their own output repo and spawn a job when a new commit is created. Similarly, the master will be able to watch a new job's input commits (which are guaranteed to exist) and begin processing data in a job when all of the job's input commits are closed.

# Data Structures
No new data structures are needed for this change, though the semantics of existing data structures change slightly:

* Commits to an output repo may not have a corresponding job yet
  * PFS creates all downstream commits in a single transaction in `FinishCommit`, but jobs are not created until PPS observes the write
* Commits do not necessary contain the result of a job
  * If a job fails, job's output commit will be a copy of the parent commit (it will be as if the commit were a dummy read-through commit)
  * This means that downstream jobs will still run. If a downstream job's input commits are all from failed jobs (which will be the case for jobs that have one input), their output commits will also, de facto, be a copy of their parents, because the job's inputs are all the same as the parent job's inputs and all cached.
  * If a failed job's output commit has no parent, the output commit will be an empty commit (*this facet of the design is still open*)

# New/Modified Operations

### Finish Commit
* This is the main operation that's changing in this refactor. Now, `FinishCommit(repo, branch) adds a commit to `branch` as well as to all repos that are provenant on `branch`.
* Note that the job that PPS creates will have to be consistent with the commit that PFS creates—otherwise, if a user calls `FinishCommit` and `UpdatePipeline` at the same time, a job and its corresponding output commit might have different provenance.
* We might ensure this by writing an [etcd revision](https://godoc.org/github.com/coreos/etcd/clientv3#WithRev) into the commit (*this facet of the design is still open*)

### Rerun Pipeline
* If a user calls `RerunPipeline` without `--reprocess`, not only will PPS create a new job to replace any failed jobs, but the failed jobs' output commits will be moved off of `master` (the pipeline's output branch), and instead, PPS will place the commit from the new job (*this is also still open. Inserting commits into the middle of a branch is not currently a supported PFS operation*)

### Delete Commit
* We can now add an operation called `DeleteCommit`. If a user deletes an input commit, PFS can simply delete all commits that are provenant on it and rewrite any deleted commits descendants to point to the deleted commit's parent.
  * If a job's output commit is deleted, the pipeline's master (which is already watching the pipeline's output repo for new commits) can simply cancel any in-process `Process()` RPCs.
  * The PPS Job object will remain in etcd (*still open*)

# Changes to PPS Job Scheduling

Pipeline masters new `watch` loop:
1. Subscribe to new commits in pipeline output repos in etcd
2. When a new commit is observed, check if the commit is to a pipeline's output repo
3. If so:
  a. determine the etcd revision at which the commit was created
  b. read the PPS Pipeline from etcd at the same revision
  c. write a corresponding PPS Job to etcd, whose output commit is the new commit.
4. Spawn k8s workers for the new job if needed, and establish GRPC client connections with them
5. Begin sending them `Process()` RPCs
6. Three cases:
  a. If the job succeeds, merge the workers' output trees into the open PFS commit, and call `FinishCommit()`
  b. If the job fails, and the output commit has a parent, copy the parent commit into the job's output commit and close it
  c. If the jobs fails and the output commit has no parent, close the output commit immediately, creating an empty commit

<sup>1</sup>Input repos are created by humans, who commit to them manually. Output repos are created by `CreatePipeline`, and their commits are created automatically.

**Comment by msteffen** [comment]
Future optimization: urrently each PPS pipeline has a `master` process running in its own pod that watches the pipeline's upstream input repos. As part of this change, pipelines may not need a dedicated master to create PPS jobs in etcd, meaning that if both this issue and https://github.com/pachyderm/pachyderm/issues/2396 are finished, we might be able to remove pipeline masters completely.

**Comment by msteffen** [comment]
I'm closing this issue because the design has been superseded by https://github.com/pachyderm/pachyderm/issues/2505
